HULKs/hulk
# Dynamic replayer

## Introduced Changes

1. Update replayer outputs when parameters are changed
2. Update replayer periodically (currently 1 Hz) to make up for the fact that we cannot react to new subscriptions from outside of communication.
3. ~~Load replay index after starting the UI~~
  3.1 ~~Show live loading progress~~
  3.2 ~~Allows scrubbing through the timeline even while parts are still being loaded~~
4. ~~Run all index building and~~ replaying in a separate thread from the UI.
  Uses `tokio::watch` to send intended time, resulting in skipped intermediate values if the replayer thread can't keep up with the requests from the UI, but with eventual consistency.

Fixes #

## ToDo / Known Issues

## Ideas for Next Iterations (Not This PR)

- Also update when new outputs are subscribed. (already soft-fixed by re-replaying at a fixed rate)
- Build recording indices in parallel instead of interleaved.

## How to Test

1. Replay something, change parameters and immediately observe the changes in the output.
2. Subscribe some new key or open image panel. After at most one second, the data should appear without having to jiggle the replayer.
3. Does not apply anymore.
4. UI should look a bit smoother on vegetables.
